controllers/admin/AdminOrdersController.php

```diff
@@ -465,47 +465,82 @@ public function postProcess()
             ShopUrl::cacheMainDomainForShop((int)$order->id_shop);
         }
 
-        /* Update shipping number */
+        /* Update shipping number and carrier */
         if (Tools::isSubmit('submitShippingNumber') && isset($order)) {
             if ($this->access('edit')) {
+                $tracking_number = Tools::getValue('shipping_tracking_number');
+                $id_carrier = Tools::getValue('shipping_carrier');
+                $old_tracking_number = $order->shipping_number;
+
                 $order_carrier = new OrderCarrier(Tools::getValue('id_order_carrier'));
                 if (!Validate::isLoadedObject($order_carrier)) {
                     $this->errors[] = $this->trans('The order carrier ID is invalid.', array(), 'Admin.OrdersCustomers.Notification');
-                } elseif (!Validate::isTrackingNumber(Tools::getValue('tracking_number'))) {
+                } elseif (!empty($tracking_number) && !Validate::isTrackingNumber($tracking_number)) {
                     $this->errors[] = $this->trans('The tracking number is incorrect.', array(), 'Admin.OrdersCustomers.Notification');
                 } else {
+                    //update carrier - ONLY if changed - then refresh shipping cost
+                    $old_id_carrier = $order_carrier->id_carrier;
+                    if (!empty($id_carrier) && $old_id_carrier != $id_carrier) {
+                        $order->id_carrier = (int) $id_carrier;
+                        $order_carrier->id_carrier = (int) $id_carrier;
+                        $order_carrier->update();
+                        $order->refreshShippingCost();
+                    }
+
+                    //load fresh order carrier because updated just before
+                    $order_carrier = new OrderCarrier((int) Tools::getValue('id_order_carrier'));
+                    
                     // update shipping number
                     // Keep these two following lines for backward compatibility, remove on 1.6 version
-                    $order->shipping_number = Tools::getValue('tracking_number');
+                    $order->shipping_number = $tracking_number;
```

Comment:
I think that this could break some tracking numbers due to problems with slashes, we have validation inside model `isTrackingNumber`